### DISCUSSION TITLE
HTML Attributes Implementation

### DISCUSSION THREAD

**Discussion by tajmone** [tajmone/Sublime-PML#37] [Syntax Elements]
@pml-lang, I need some help implementing [HTML attributes] support.

I'm not quite sure about various choices here, because this element can be found in many forms. So, here are some questions:

1. In terms of RegEx, for the pattern following the `html_` prefix, what characters set should I expect? Just lowercase or also uppercase? Underscore and/or hyphens?
2. The full list of all the possible HTML attributes is too long to cover each one of them with a specific context, so for uncommon attributes we could use just a generic scope.
3. Some common HTML attributes like `style`, `class`, etc., might be worth capturing with a specific context. E.g. for `style` some CSS highlighting might be used. Which are the attributes that are worth covering with a dedicated context?

Obviously, if the package is able to track some of these attributes it could also mean that in the future it will allow dedicated features for working upon them — e.g. refactoring class names with `html_class`, etc. Hence my questions, since I'm trying to understand the full potential of PML's support for [HTML attributes] and what is your vision about this element.

Do you have any advice an suggestions in this respect?

-------------------------------------------------------------------------------

## Reference Links

- [Mozilla » HTML attribute reference]
- [W3Schools » HTML Attribute Reference]
- [BitDegree » Using HTML Attributes]

[HTML Attributes]: https://www.pml-lang.dev/docs/user_manual/index.html#HTML_attributes

[Mozilla » HTML attribute reference]: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes
[W3Schools » HTML Attribute Reference]: https://www.w3schools.com/tags/ref_attributes.asp

[BitDegree » Using HTML Attributes]: https://www.bitdegree.org/learn/html-attributes "Using HTML Attributes: List of the Most Common Attributes Explained"

**Comment by pml-lang** [comment]
> In terms of RegEx, for the pattern following the html_ prefix, what characters set should I expect? Just lowercase or also uppercase? Underscore and/or hyphens?

PMLC works like this:

- Attribute names must respect the [PDML rule for node names](https://pdml-lang.github.io/docs/basic-specification/index.html#node_name), which is also applied for _attribute_ names.
- If an attribute name starts with `html_` (must be lowercase), then the `html_` prefix is removed, and the rest is considered to be a valid HTML attribute. PMLC does not check if it is indeed a _valid_ HTML attribute. Therefore: Yes, lowercase/uppercase letters, underscores and hyphens are all valid characters in PMLC. Even digits and dots are allowed (according to the `[a-zA-Z_][a-zA-Z0-9_\.-]*` regex rule), although (as far as I know) those are not used in HTML attributes.

> The full list of all the possible HTML attributes is too long to cover each one of them with a specific context, so for uncommon attributes we could use just a generic scope.

Yes, as said already, PMLC does not handle the full list of all the possible HTML attributes.

Note: I've just added the following sentence at the end of chapter [HTML attributes](https://www.pml-lang.dev/docs/user_manual/index.html#HTML_attributes) (will be included in version 3.0.0): "PMLC does not check if an HTML attribute is valid."

> Some common HTML attributes like style, class, etc., might be worth capturing with a specific context. E.g. for style some CSS highlighting might be used. Which are the attributes that are worth covering with a dedicated context?

Yes, I think that `style` and `class` are those that might be worth capturing with a specific context.
